victor/src/layout/mod.rs

```diff
@@ -160,14 +249,253 @@ fn same_formatting_context_block(
+    let computed_margin = style
+        .margin()
+        .map(|v| v.non_auto().map(|v| v.percentage_relative_to(cbis)));
+
+    let computed_inline_size = box_size
+        .inline
+        .non_auto()
+        .map(|v| v.percentage_relative_to(cbis));
+    let computed_block_size = box_size.block.non_auto().and_then(|b| match b {
+        LengthOrPercentage::Length(l) => Some(l),
+        LengthOrPercentage::Percentage(p) => {
+            absolute_containing_block.block_size.map(|cbbs| cbbs * p)
+        }
+    });
+
+    struct Solution {
+        margin_start: Length,
+        margin_end: Length,
+        strategy: Strategy,
+    }
+
+    enum Strategy {
+        FromStart {
+            start: Option<Length>,
+            size: Option<Length>,
+        },
+        FromEnd {
+            end: Length,
+        },
+    }
+
+    fn solve_axis(
+        containing_block_inline_size: Length,
+        box_start: Option<Length>,
+        box_end: Option<Length>,
+        computed_margin_start: Option<Length>,
+        computed_margin_end: Option<Length>,
+        solve_margins: impl FnOnce(Length) -> (Length, Length),
+        padding_border_sum: Length,
+        size: Option<Length>,
+    ) -> Solution {
+        let cbis_minus_pb = containing_block_inline_size - padding_border_sum;
+        let zero = Length::zero();
+
+        let mut margin_start = computed_margin_start.unwrap_or(zero);
+        let mut margin_end = computed_margin_end.unwrap_or(zero);
+
+        let strategy = match (box_start, size, box_end) {
+            (start, size, None) => Strategy::FromStart { start, size },
+            (Some(start), Some(size), Some(end)) => {
+                let margins = cbis_minus_pb - start - size - end;
+
+                match (computed_margin_start, computed_margin_end) {
+                    (None, None) => {
+                        let (s, e) = solve_margins(margins);
+                        margin_start = s;
+                        margin_end = e;
+                    }
+                    (None, Some(end)) => {
+                        margin_start = margins - end;
+                        margin_end = end;
+                    }
+                    (Some(start), _) => {
+                        margin_start = start;
+                        margin_end = margins - start;
+                    }
+                }
+
+                Strategy::FromStart {
+                    start: Some(start),
+                    size: Some(size),
+                }
+            }
+            (None, None, Some(end)) => Strategy::FromEnd { end },
+            (None, Some(size), Some(end)) => {
+                let start = cbis_minus_pb - size - end - margin_start - margin_end;
+                Strategy::FromStart {
+                    start: Some(start),
+                    size: Some(size),
+                }
+            }
+            (Some(start), None, Some(end)) => {
+                // FIXME(nox): Wait, what happens when that is negative?
```

Comment:
This might be a spec bug? Pretty sure the content width (and height) is supposed to be non-negative. Could you file it w3c/csswg-drafts?

Reply:
https://drafts.csswg.org/css2/visudet.html#propdef-width
> Negative values for 'width' are illegal. 

But that’s specified values.